### PR TITLE
README: Clarify cookie security implications

### PR DESCRIPTION
**Pull request by wmertens**
This explanation would have prevented me from opening #195

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +54,12 @@ any of the following keys:
 Determines if the token secret for the user should be stored in a cookie
 or in `req.session`. Defaults to `false`.
 
+This implements the [Double Submit Cookie Pattern](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie) and is
+easy to implement, but you should be sure that the caveats do not apply to you.
+
+In particular, the client still needs to send the CSRF token in a separate
```
Comment (dougwilson) [inline README.md]: What does this part mean? I didn't see that on the OWASP page, and not sure what that means in this module, either, as this module does not include the token in a cookie (just the secret is in the cookie). As the author, I'm not even sure what this means in order to follow, so unsure other users will too, which is why I'm asking if it can be clarified.

File: README.md [hunk 2]
```diff
@@ -54,6 +54,12 @@ any of the following keys:
 Determines if the token secret for the user should be stored in a cookie
 or in `req.session`. Defaults to `false`.
 
+This implements the [Double Submit Cookie Pattern](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie) and is
```
Comment (dougwilson) [inline README.md]: Sorry, I guess re-reading this part also needs some clarity. It says "This implements the double submit cookie" here, but what does "This" refer to? If it refers to "This" being when the option is set to true (and thus not to false), maybe clarify or move lower to after the description of setting to true is my thought.

File: README.md (hunk 1 above)
Comment (wmertens) [reply README.md]: I saw a few issues about people adding middleware for sending the token as a cookie, thereby completely invalidating the concept of CSRF.

I wanted to warn against that. How about

> In particular, the CSRF token itself should not be put in a cookie, since that would remove the CSRF protection. Instead, it should be passed to the client so the client can present it to validate the request.

?

Comment (dougwilson) [reply README.md]: How does putting into a cookie compromise csrf protection, specifically? This module does not read the value back from a cookie, only from a separate request parameter. Even if the user puts in in a cookie, this module will not honor it.

Comment (wmertens) [reply README.md]: ah - I misread the configuration, I was under the impression that you could provide the token in a cookie and the middleware would read it from there as well. I'll remove this paragraph then.

Comment (dougwilson) [reply README.md]: Gotcha. This only stores the secret in the cookie. The reason why you see people putting the token in a cookie is because csrf plugins for things like angular will read the token from there client side, but then still include it as a request parameter on submit.

**Comment by wesselvdv** [comment]
Apologies, but I don't see how csurf applies the double-submit technique as is being described in this merge request. 

Quote from OWASP:
```
This technique is easy to implement and is stateless. In this technique, we send a random value in both a cookie and as a request parameter, with the server verifying if the cookie value and request value match.
```

I fail to see where csurf is comparing the contents of the `_csrf` cookie, and that of another request parameter? Either I am looking at the wrong things, or csurf only validates either one of them. Thus doesn't implement the Double Submit Cookie technique.

**Comment by dougwilson** [comment]
Hi @wesselvdv sorry if the code is hard to follow. You're welcome to use a debugger to walk through the code to follow the flow if it help you verify, but it does indeed compare them. You can try it yourself by not submitting the value or altering the cookie value and then submitting the same request value, etc.

The comparison between the two values is occurring right here:

https://github.com/expressjs/csurf/blob/248112a42f36fc9a84a71b0f5d383a1e03813f54/index.js#L111

The `secret` variable is the value from the cookie and the `value(req)` function call gets the value from the request to compare against.

**Comment by wesselvdv** [comment]
> Hi @wesselvdv sorry if the code is hard to follow. You're welcome to use a debugger to walk through the code to follow the flow if it help you verify, but it does indeed compare them. You can try it yourself by not submitting the value or altering the cookie value and then submitting the same request value, etc.
> 
> The comparison between the two values is occurring right here:
> 
> https://github.com/expressjs/csurf/blob/248112a42f36fc9a84a71b0f5d383a1e03813f54/index.js#L111
> 
> The `secret` variable is the value from the cookie and the `value(req)` function call gets the value from the request to compare against.

Awesome! I am glad it does do this, ignore my previous comment/rant.
